mm2src/coins/lightning/ln_events.rs

```diff
@@ -432,12 +458,23 @@ impl LightningEventHandler {
             },
         };
 
+        let claiming_txid = claiming_tx.txid();
+        let tx_hex = serialize_hex(&claiming_tx);
+        if let Err(e) = tokio::task::block_in_place(move || self.platform.coin.send_raw_tx(&tx_hex).wait()) {
```

Comment:
> The BackgroundProcessor loop is spawned inside the start function https://github.com/lightningdevkit/rust-lightning/blob/d26640446f54342184fdfee9cc27efe9ce8f290e/lightning-background-processor/src/lib.rs#L208 which we use here 

As I can see, this is started in a separate thread. How come that 
> channel opening because ping/pong messages weren't sent/received in time due to blocking

Is there a chance that these somehow reach tokio async runtime, blocking it? `block_in_place` simply blocks the current thread if not running on tokio runtime:
https://github.com/tokio-rs/tokio/blob/master/tokio/src/runtime/thread_pool/worker.rs#L316
https://github.com/tokio-rs/tokio/blob/master/tokio/src/runtime/thread_pool/worker.rs#L355

So for blocking fns running in a separate thread, `block_in_place` should have no effect, but it does.

Reply:
Apparently `block_in_place` was allowed to be used outside the tokio runtime in this PR https://github.com/tokio-rs/tokio/pull/2410, I am still trying to figure out how this all works :)

Reply:
> Apparently block_in_place was allowed to be used outside the tokio runtime in this PR https://github.com/tokio-rs/tokio/pull/2410, I am still trying to figure out how this works :)

Yes, it simply blocks the current thread if not executed on the async runtime. But if this is required, then I assume it somehow reaches tokio runtime, which shouldn't happen. I'm mostly curious why it happens than considering this a real problem, though 🙂

Reply:
> Yes, it simply blocks the current thread if not executed on the async runtime. 

After debugging, that's correct. This should not be a problem in the `BackgroundProcessor` spawned loop, at most `get_est_sat_per_1000_weight` is called once every loop and will take at most 60 seconds if it timed out, but maybe having an async loop in our code that calculates fees every few seconds and saving it to a variable that gets used in `get_est_sat_per_1000_weight` straight away might be a good solution. What do you think @artemii235? I will elaborate more on this, but for now based on the debugging I guess this https://github.com/KomodoPlatform/atomicDEX-API/blob/3fd5e14088ff9523873eb7ddeb4ca3b7c56ed6b4/mm2src/coins/lightning/ln_events.rs#L463 needs to use `spawn` instead since `block_in_place` does nothing as you have noticed.

As for the old problems I mentioned above about peer disconnections before using `block_in_place` in `get_est_sat_per_1000_weight` https://github.com/KomodoPlatform/atomicDEX-API/pull/1339#discussion_r924620626, this was due to a different issue related to inbound connections. Peer disconnections happened when I was running two mm2 lightning nodes and the peers always disconnected when opening a channel between them, due to `get_est_sat_per_1000_weight` blocking ping/pong messages to be received in time by the inbound node. https://github.com/KomodoPlatform/atomicDEX-API/blob/3fd5e14088ff9523873eb7ddeb4ca3b7c56ed6b4/mm2src/coins/lightning/ln_p2p.rs#L155-L157 Since [rust-lightning v0.0.105](https://github.com/lightningdevkit/rust-lightning/pull/1281) `get_est_sat_per_1000_weight` for inbound channels is called by `accept_inbound_channel` https://github.com/KomodoPlatform/atomicDEX-API/blob/3fd5e14088ff9523873eb7ddeb4ca3b7c56ed6b4/mm2src/coins/lightning/ln_events.rs#L533 due to having a new `OpenChannelRequest` event, this would have also lead to blocking if `block_in_place` weren't used.

I also found other problems related to using `block_in_place` inside `get_est_sat_per_1000_weight` after debugging our tokio runtime using [tokio-console](https://github.com/tokio-rs/console), as the task started by `close_channel` https://github.com/KomodoPlatform/atomicDEX-API/blob/3fd5e14088ff9523873eb7ddeb4ca3b7c56ed6b4/mm2src/coins/lightning.rs#L1484 doesn't end as shown below

<img width="1716" alt="Screen Shot 2022-07-22 at 7 09 29 PM" src="https://user-images.githubusercontent.com/39480341/180788252-35bcde8c-3692-460e-927d-e4ca34aa4472.png">

<img width="1594" alt="Screen Shot 2022-07-22 at 7 10 06 PM" src="https://user-images.githubusercontent.com/39480341/180788876-54076aed-91ae-45b4-9470-4670437ec2b6.png">

This happens due to using `block_in_place` from inside the async function straight away, and it doesn't happen when I wrapped the call inside `spawn_blocking` as in the `open_channel` call https://github.com/KomodoPlatform/atomicDEX-API/blob/3fd5e14088ff9523873eb7ddeb4ca3b7c56ed6b4/mm2src/coins/lightning.rs#L881-L886 where the task finishes as soon as the `open_channel` call is done.

In conclusion, having a variable that stores the fees every few seconds and using it inside `get_est_sat_per_1000_weight` as stated above can solve a lot of problems, and only one task will be spawned for this.

Reply:
Thanks for the comprehensive research! 🙂 

> In conclusion, having a variable that stores the fees every few seconds and using it inside get_est_sat_per_1000_weight as stated above can solve a lot of problems, and only one task will be spawned for this.

Yeah, seems a solution 🙂 A small drawback is you will likely have to call `rpc_client().estimate_fee_sat` multiple times per each iteration for each possible `ConfirmationTarget` variant. Does it have to be called every few seconds? How frequently and significantly fee changes on BTC network?

Reply:
> How frequently and significantly fee changes on BTC network?

By quickly checking the bitcoin code, The fee seems to be the max of `min_mempool_feerate`, `min_relay_feerate` https://github.com/bitcoin/bitcoin/blob/6dc3084eec912cf2abfe18d1c05655defaa45e20/src/rpc/fees.cpp#L92
And the `min_mempool_feerate` gets recalculated every 10 seconds or more https://github.com/bitcoin/bitcoin/blob/948f5ba6363fcc64f95fed3f04dbda3d50d61827/src/txmempool.cpp#L1110 and it depends on the size of the mempool, so if a surge of transactions arrive at once it can change significantly on the next recalculation.

> Does it have to be called every few seconds? 

Of Course not, but I am not sure what the best interval can be, 10 seconds is the safest but I think it can be increased. Also this might be complicated but maybe there is a calculation for the target that can be done on our code while requesting the fee only one time, I will look into it :)

Reply:
PS you can do it on the next iteration 🙂

Reply:
> Also this might be complicated but maybe there is a calculation for the ConfirmationTarget that can be done on our code while requesting the fee only one time instead of 3 times (one per every ConfirmationTarget)

Unfortunately, this can't be done since a sample of the txs in the mempool is used in calculating the fee for a given `ConfirmationTarget`
https://github.com/bitcoin/bitcoin/blob/6dc3084eec912cf2abfe18d1c05655defaa45e20/src/rpc/fees.cpp#L88
https://github.com/bitcoin/bitcoin/blob/6dc3084eec912cf2abfe18d1c05655defaa45e20/src/policy/fees.cpp#L769
https://github.com/bitcoin/bitcoin/blob/6dc3084eec912cf2abfe18d1c05655defaa45e20/src/policy/fees.cpp#L269-L270
This will also lead to the fee for a given target changing in less than 10 seconds depending on the txs in the mempool

There is no alternative to an estimate fee request for every `ConfirmationTarget` :(

Reply:
> PS you can do it on the next iteration 🙂

Added this to the issue's checklist https://github.com/KomodoPlatform/atomicDEX-API/issues/1045#issue-971642359